REGRESSION(274481@main): crash when trying to open web inspector, `RELEASE_ASSERT(!m_inStyleRecalc)` hit under `Document::resolveStyle()`

dcrousso opened the pull request:
#### 0683db19e202d5882d533a6049780ceb51bb1e76
<pre>
REGRESSION(274481@main): crash when trying to open web inspector, `RELEASE_ASSERT(!m_inStyleRecalc)` hit under `Document::resolveStyle()`
<a href="https://bugs.webkit.org/show_bug.cgi?id=286534">https://bugs.webkit.org/show_bug.cgi?id=286534</a>
&lt;<a href="https://rdar.apple.com/problem/143769557">rdar://problem/143769557</a>&gt;

Reviewed by BJ Burg.

Instead of pushing animation effect changes to the frontend every time they happen, have the frontend request the new animation effect only when it needs it.

This avoids the unnecessary work of recalculating and redisplaying the animation effect data when the Graphics Tab is not visible (as well as fixing the bug by not doing work during layout).

For future reference, an alternative fix could&apos;ve been to modify `InspectorAnimationAgent::enable` such that it delays calling `InspectorAnimationAgent::bindAnimation` by at least `0_s` (similar to what `InspectorAnimationAgent::didCreateWebAnimation` already does) so that it can never happen during layout.

* Source/JavaScriptCore/inspector/protocol/Animation.json:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::requestEffect): Added.
(WebCore::InspectorAnimationAgent::requestEffectTarget):
(WebCore::InspectorAnimationAgent::didChangeWebAnimationEffectTiming):
(WebCore::InspectorAnimationAgent::didChangeWebAnimationEffectTarget):
(WebCore::InspectorAnimationAgent::bindAnimation):
* Source/WebInspectorUI/UserInterface/Protocol/AnimationObserver.js:
(WI.AnimationObserver.prototype.effectChanged):
* Source/WebInspectorUI/UserInterface/Controllers/AnimationManager.js:
(WI.AnimationManager.prototype.effectChanged):
* Source/WebInspectorUI/UserInterface/Models/Animation.js:
(WI.Animation):
(WI.Animation.fromPayload):
(WI.Animation.prototype.get startDelay):
(WI.Animation.prototype.get endDelay):
(WI.Animation.prototype.get iterationCount):
(WI.Animation.prototype.get iterationStart):
(WI.Animation.prototype.get iterationDuration):
(WI.Animation.prototype.get timingFunction):
(WI.Animation.prototype.get playbackDirection):
(WI.Animation.prototype.get fillMode):
(WI.Animation.prototype.get keyframes):
(WI.Animation.prototype.ensureEffect): Added.
(WI.Animation.prototype.effectChanged):
(WI.Animation.prototype._updateEffect):

* Source/WebInspectorUI/UserInterface/Views/AnimationContentView.js:
(WI.AnimationContentView.prototype.async _refreshPreview):
* Source/WebInspectorUI/UserInterface/Views/AnimationDetailsSidebarPanel.js:
(WI.AnimationDetailsSidebarPanel.prototype.layout):
(WI.AnimationDetailsSidebarPanel.prototype.async _refreshEffectSection):
(WI.AnimationDetailsSidebarPanel.prototype.attached): Deleted.
Ensure that the effect is updated before attempting to grab any values from it.
Drive-by: simplify how `CodeMirror.prototype.refresh` is called after updating the effect.

* LayoutTests/inspector/animation/resources/lifecycle-utilities.js:
* LayoutTests/inspector/animation/effectChanged.html:
* LayoutTests/inspector/animation/effectChanged-expected.txt:

Canonical link: <a href="https://commits.webkit.org/292391@main">https://commits.webkit.org/292391@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0b84fb848674d2730d8b9d29350020d49929045

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95937 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15551 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5497 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/100998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/46446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97982 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15846 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23983 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/100998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/46446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98940 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/11862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86655 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/100998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/11596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4406 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/45781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88610 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4508 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103025 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94558 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23004 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23255 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82681 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81552 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20443 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26141 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22967 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/118035 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/22626 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/26106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24367 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebCore/inspector/agents/InspectorAnimationAgent.h:
```diff
@@ -98,20 +99,23 @@ class InspectorAnimationAgent final : public InspectorAgentBase, public Inspecto
     WeakRef<Page> m_inspectedPage;
 
-    MemoryCompactRobinHoodHashMap<String, WebAnimation*> m_animationIdMap;
+    MemoryCompactRobinHoodHashMap<Inspector::Protocol::Animation::AnimationId, WebAnimation*> m_animationIdMap;
```
I know you didn't change this, but what owns the memory of these WebAnimations? Storing raw pointers seems like the wrong direction.

dcrousso replied:
this is how Web Inspector has done it for years.  we dont want to keep anything alive as that would kinda defeat the purpose of being a (somewhat) passive observer.  instead, we use `InspectorInstrumentation` to get notified right after creation and before destruction to add/remove from these maps

nowadays we probably should use something like `WeakRef`, even though these will only exist within the created->destroyed lifecycle (via `InspectorInstrumentation`)

burg commented on Source/WebInspectorUI/UserInterface/Models/Animation.js:
```diff
@@ -282,7 +332,11 @@ WI.Animation = class Animation extends WI.Object
 
-        this.dispatchEventToListeners(WI.Animation.Event.EffectChanged);
+        if (this._ensureEffectCallbacks) {
+            for (let requestEffectCallback of this._ensureEffectCallbacks)
```
Is it impossible that _ensureEffectCallbacks would be mutated by requestEffectCallback? If not, please iterate over a copy.

dcrousso replied:
yes by this point `_effect` will have been set, meaning that `Animation.prototype.ensureEffect` will immediately invoke the callback or return a `Promise` instead of modifying `_ensureEffectCallbacks` (i.e. the only way for it to be modified is if `_effect` is not set yet, and this code runs when that eventually happens)

dcrousso committed:
REGRESSION(274481@main): crash when trying to open web inspector, `RELEASE_ASSERT(!m_inStyleRecalc)` hit under `Document::resolveStyle()`
https://bugs.webkit.org/show_bug.cgi?id=286534
<rdar://problem/143769557>

Reviewed by BJ Burg.

Instead of pushing animation effect changes to the frontend every time they happen, have the frontend request the new animation effect only when it needs it.

This avoids the unnecessary work of recalculating and redisplaying the animation effect data when the Graphics Tab is not visible (as well as fixing the bug by not doing work during layout).

For future reference, an alternative fix could've been to modify `InspectorAnimationAgent::enable` such that it delays calling `InspectorAnimationAgent::bindAnimation` by at least `0_s` (similar to what `InspectorAnimationAgent::didCreateWebAnimation` already does) so that it can never happen during layout.

* Source/JavaScriptCore/inspector/protocol/Animation.json:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::requestEffect): Added.
(WebCore::InspectorAnimationAgent::requestEffectTarget):
(WebCore::InspectorAnimationAgent::didChangeWebAnimationEffectTiming):
(WebCore::InspectorAnimationAgent::didChangeWebAnimationEffectTarget):
(WebCore::InspectorAnimationAgent::bindAnimation):
* Source/WebInspectorUI/UserInterface/Protocol/AnimationObserver.js:
(WI.AnimationObserver.prototype.effectChanged):
* Source/WebInspectorUI/UserInterface/Controllers/AnimationManager.js:
(WI.AnimationManager.prototype.effectChanged):
* Source/WebInspectorUI/UserInterface/Models/Animation.js:
(WI.Animation):
(WI.Animation.fromPayload):
(WI.Animation.prototype.get startDelay):
(WI.Animation.prototype.get endDelay):
(WI.Animation.prototype.get iterationCount):
(WI.Animation.prototype.get iterationStart):
(WI.Animation.prototype.get iterationDuration):
(WI.Animation.prototype.get timingFunction):
(WI.Animation.prototype.get playbackDirection):
(WI.Animation.prototype.get fillMode):
(WI.Animation.prototype.get keyframes):
(WI.Animation.prototype.ensureEffect): Added.
(WI.Animation.prototype.effectChanged):
(WI.Animation.prototype._updateEffect):

* Source/WebInspectorUI/UserInterface/Views/AnimationContentView.js:
(WI.AnimationContentView.prototype.async _refreshPreview):
* Source/WebInspectorUI/UserInterface/Views/AnimationDetailsSidebarPanel.js:
(WI.AnimationDetailsSidebarPanel.prototype.layout):
(WI.AnimationDetailsSidebarPanel.prototype.async _refreshEffectSection):
(WI.AnimationDetailsSidebarPanel.prototype.attached): Deleted.
Ensure that the effect is updated before attempting to grab any values from it.
Drive-by: simplify how `CodeMirror.prototype.refresh` is called after updating the effect.

* LayoutTests/inspector/animation/resources/lifecycle-utilities.js:
* LayoutTests/inspector/animation/effectChanged.html:
* LayoutTests/inspector/animation/effectChanged-expected.txt:

Canonical link: https://commits.webkit.org/292391@main
